### PR TITLE
WIP: Add a field non-preempting to PriorityClass

### PR DESCRIPTION
**Pull request by tanshanshan**
**What this PR does / why we need it**:
as #67671 ,add a field non-preempting to PriorityClass,

A Pod has preempting priority, unless its PriorityClass is found and the non-preempting field of the PriorityClass is set to false. 
**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
Add a field non-preempting to PriorityClass 
```

### CODE REVIEW

**Comment by k82cn** [comment]
please separate the feature into several commits :)

File: pkg/scheduler/util/utils.go [hunk 1]
```diff
@@ -177,6 +180,18 @@ func GetPodPriority(pod *v1.Pod) int32 {
...
 
+// GetPodNonPreempting return the NonPreempting of the priorityClass.
+func GetPodNonPreempting(pod *v1.Pod, kubeClient clientset.Interface) bool {
+	if pod.Spec.Priority == nil {
+		return false
+	}
+	if kubeClient == nil {
+		return false
+	}
+	priorityClass, _ := kubeClient.SchedulingV1beta1().PriorityClasses().Get(pod.Name, metav1.GetOptions{})
```
Comment (yastij) [inline pkg/scheduler/util/utils.go]: use `pod.PriorityClassName` instead ?

File: pkg/scheduler/scheduler_test.go [hunk 2]
```diff
@@ -134,7 +135,7 @@ func (es mockScheduler) Prioritizers() []algorithm.PriorityConfig {
 	return nil
 }
 
-func (es mockScheduler) Preempt(pod *v1.Pod, nodeLister algorithm.NodeLister, scheduleErr error) (*v1.Node, []*v1.Pod, []*v1.Pod, error) {
+func (es mockScheduler) Preempt(pod *v1.Pod, nodeLister algorithm.NodeLister, kubeClient clientset.Interface, scheduleErr error) (*v1.Node, []*v1.Pod, []*v1.Pod, error) {
```
Comment (yastij) [inline pkg/scheduler/scheduler_test.go]: why pass the kubeClient and not adding it to `mockScheduler` ?

**Review by yastij — CHANGES_REQUESTED** [review]
left a few comments. @bsalamat can you take a look when you have some bandwidth ?

File: pkg/scheduler/util/utils.go [hunk 3]
```diff
@@ -188,7 +188,7 @@ func GetPodNonPreempting(pod *v1.Pod, kubeClient clientset.Interface) bool {
 	if kubeClient == nil {
 		return false
 	}
-	priorityClass, _ := kubeClient.SchedulingV1beta1().PriorityClasses().Get(pod.Name, metav1.GetOptions{})
+	priorityClass, _ := kubeClient.SchedulingV1beta1().PriorityClasses().Get(pod.Spec.PriorityClassName, metav1.GetOptions{})
```
Comment (yastij) [inline pkg/scheduler/util/utils.go]: I'd really prefer having a cache here that gets updated using `informers`. But I'll defer to @bsalamat and @k82cn

File: pkg/apis/scheduling/types.go [hunk 4]
```diff
@@ -64,6 +64,12 @@ type PriorityClass struct {
 	// when this priority class should be used.
 	// +optional
 	Description string
+
+	// NonPreempting specifies whether the pod which used this PriorityClass can start a preempting process.
```
Comment (bsalamat) [inline pkg/apis/scheduling/types.go]: Given that a pod cannot start a preemption process, let's change the comment to:
```
// nonPreempting specifies whether a pod with this PriorityClass could trigger a preemption process.
// If this field is unset, the PriorityClass is considered a preempting class by default.
```

File: pkg/kubectl/cmd/create/create_priorityclass.go [hunk 5]
```diff
@@ -28,7 +28,7 @@ import (
 
 var (
 	pcLong = templates.LongDesc(i18n.T(`
-		Create a priorityclass with the specified name, value, globalDefault and description`))
+		Create a priorityclass with the specified name, value, globalDefault nonPreempting,and description`))
```
Comment (bsalamat) [inline pkg/kubectl/cmd/create/create_priorityclass.go]: s/globalDefault nonPreempting,and/globalDefault, nonPreempting, and/

File: pkg/kubectl/cmd/create/create_priorityclass.go [hunk 6]
```diff
@@ -70,6 +70,8 @@ func NewCmdCreatePriorityClass(f cmdutil.Factory, ioStreams genericclioptions.IO
 	cmd.Flags().Int32("value", 0, i18n.T("the value of this priority class."))
 	cmd.Flags().Bool("global-default", false, i18n.T("global-default specifies whether this PriorityClass should be considered as the default priority."))
 	cmd.Flags().String("description", "", i18n.T("description is an arbitrary string that usually provides guidelines on when this priority class should be used."))
+	cmd.Flags().Bool("non-preempting", false, i18n.T("non-preempting specifies whether the pod which used this PriorityClass can start a preempting process."))
```
Comment (bsalamat) [inline pkg/kubectl/cmd/create/create_priorityclass.go]: Please change description similar to the comment for the "nonPreempting" field.

File: pkg/scheduler/util/utils.go (hunk 3 above)
Comment (bsalamat) [reply pkg/scheduler/util/utils.go]: Yes, getting PriorityClasses from the API server in critical path of scheduling could have major performance implications. We should use informers and cache them.

**Review by bsalamat — COMMENTED** [review]
@tanshanshan Thanks and sorry for the delay in reviewing your PR.
Two issues stand out in this PR:
1. We should use informers instead of getting PriorityClasses from the API server (more details in the comments).
2. We need more tests to verify that:
   a. Creation of PriorityCalsses with non-preempting field set to true works.
   b. Pods with non-preempting priority do not preempt other pods despite their priority being higher.

**Comment by tanshanshan** [comment]
@bsalamat  
 I will finish it ASAP.

**Comment by k82cn** [comment]
> By the way , testcase should be in this pr, or another pr is ok?

In this PR by separated commits :)

**Comment by k82cn** [comment]
copy from slack: btw, we need to clarify the behaviour on PriorityClass modification, .e.g. updated.

**Comment by bsalamat** [comment]
@tanshanshan there are build errors. Please take a look.

**Comment by tanshanshan** [comment]
> @tanshanshan there are build errors. Please take a look.

I got a typecheck error:
ERROR(all) pkg/scheduler/core/generic_scheduler_test.go:1510:47: too few arguments in call to NewGenericScheduler
exit status 1

but  I use make verify WHAT=typecheck in my local is ok

in my code  pkg/scheduler/core/generic_scheduler_test.go L1501
https://github.com/kubernetes/kubernetes/pull/67742/files#diff-718c5036402a4013245b26f23dc1d824R1501

the number of arguments  is right.

I retest many times, but it is also not ok

**Comment by k82cn** [comment]
> but I use make verify WHAT=typecheck in my local is ok

try this one:

```
make test WHAT=./pkg/scheduler/core/
```

File: staging/src/k8s.io/api/scheduling/v1alpha1/types.go [hunk 7]
```diff
@@ -49,6 +49,11 @@ type PriorityClass struct {
 	// when this priority class should be used.
 	// +optional
 	Description string `json:"description,omitempty" protobuf:"bytes,4,opt,name=description"`
+
+	// nonPreempting specifies whether a pod with this PriorityClass could trigger a preemption process.
+	// If this field is unset, the PriorityClass is considered a preempting class by default.
+	// +optional
+	NonPreempting bool `json:"nonPreempting,omitempty" protobuf:"bytes,5,opt,name=nonPreempting"`
```
Comment (k82cn) [inline staging/src/k8s.io/api/scheduling/v1alpha1/types.go]: do we still need to support v1alpha1 for this feature?

Comment (tanshanshan) [reply staging/src/k8s.io/api/scheduling/v1alpha1/types.go]: also remove the field  NonPreempting  from  pkg/apis/scheduling/types.go?

**Comment by tanshanshan** [comment]
I add field NonPreempting in  pkg/apis/scheduling/types.go and staging/src/k8s.io/api/scheduling/v1beta1/types.go

but verify failed ,

```
E0929 04:31:07.190206    6019 conversion.go:667] Warning: could not find nor generate a final Conversion function for k8s.io/kubernetes/pkg/apis/scheduling.PriorityClass -> k8s.io/api/scheduling/v1alpha1.PriorityClass
E0929 04:31:07.190268    6019 conversion.go:668]   the following fields need manual conversion:
E0929 04:31:07.190275    6019 conversion.go:670]       - NonPreempting
2018/09/29 04:31:15 Code for OpenAPI definitions generated
# k8s.io/kubernetes/pkg/apis/scheduling/v1alpha1
pkg/apis/scheduling/v1alpha1/zz_generated.conversion.go:43:10: undefined: Convert_scheduling_PriorityClass_To_v1alpha1_PriorityClass
pkg/apis/scheduling/v1alpha1/zz_generated.conversion.go:109:14: undefined: Convert_scheduling_PriorityClass_To_v1alpha1_PriorityClass
```
but in this file func Convert_scheduling_PriorityClass_To_v1alpha1_PriorityClass exists,
so how to fix it 
I try to add the field to staging/src/k8s.io/api/scheduling/v1alpha1/types.go ,test will be ok.


@bsalamat

**Comment by bsalamat** [comment]
@tanshanshan Since you have added new fields that didn't exist in v1alpha1, you should provide a "defaults.go" file for the version. Please [see here](https://github.com/kubernetes/community/blob/master/contributors/devel/api_changes.md#edit-defaultsgo) for more information. In this case, it will be a simple file. You can refer to [the storage API defaults.go](https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/storage/v1beta1/defaults.go) as an example.

**Comment by tanshanshan** [comment]
I am in a holiday , and I will fix it in the next week. 
Thanks. @bsalamat

**Comment by bsalamat** [comment]
ping @tanshanshan 
could you please take a look at the errors and fix them? Looks like you are missing those convert functions.

**Comment by tanshanshan** [comment]
I add defaults.go in pkg/apis/scheduling/v1alpha1/defaults.go and pkg/apis/scheduling/v1beta1/defaults.go

and this field is bool, so I think it is not give a default value.


but this error also exists 

> pkg/apis/scheduling/v1alpha1/zz_generated.conversion.go:43:10: undefined: Convert_scheduling_PriorityClass_To_v1alpha1_PriorityClass
> pkg/apis/scheduling/v1alpha1/zz_generated.conversion.go:109:14: undefined: Convert_scheduling_PriorityClass_To_v1alpha1_PriorityClass

but Convert_scheduling_PriorityClass_To_v1alpha1_PriorityClass  in pkg/apis/scheduling/v1alpha1/zz_generated.conversion.go exists, 
so I  am confused.

**Comment by tanshanshan** [comment]
I add 

```
func SetDefaults_PriorityClass(obj *schedulingv1beta1.PriorityClass) {
	obj.NonPreempting = false
}
```

to pkg/apis/scheduling/v1beta1/defaults.go

**Comment by bsalamat** [comment]
@tanshanshan Do you have the time to check the test failures and fix them?

**Comment by tanshanshan** [comment]
> @tanshanshan Do you have the time to check the test failures and fix them?

I see the error about bazel test, but I  do not understand how to fix it.

**Comment by k82cn** [comment]
@tanshanshan , are you still working on this ? If not, I'd like to raise in community to let others pick up :)

**Comment by tanshanshan** [comment]
yes ,  I  am working ,but  I  do not know how to deal with  bazel-test  failed

**Pull request closed without merging**